wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -59,12 +67,39 @@ applyBlock pw@PassiveWallet{..} b = do
     -- apply block to all Accounts in all Wallets
     mConfirmed <- update' _wallets $ ApplyBlock k ctxt blocksByAccount
     case mConfirmed of
-      Left  err       -> return $ Left err
+      Left  (ApplyBlockNotSuccessor curCtx cpCtx) -> runExceptT (backfill (Just curCtx) cpCtx)
+      Left err -> return $ Left err
       Right confirmed -> do
         modifyMVar_ _walletSubmission $ return . Submission.remPending confirmed
         mapM_ (putTxMeta _walletMeta) metas
         return $ Right ()
 
+  where
+      -- Backfill blocks that were missing between the given block and the wallet's most recent
+      -- checkpoint. 'backfill src tgt' will apply all blocks in order, starting from the
+      -- successor of tgt until src is reached. 'Nothing' is used to represent the genesis block.
+      backfill :: Maybe BlockContext -> Maybe BlockContext -> ExceptT ApplyBlockFailed IO ()
+      backfill Nothing Nothing   = return ()
+      backfill Nothing (Just cp) = throwError (CouldNotReachCheckpoint cp)
+      backfill (Just cur) tgt    =
+        if (Just (cur ^. bcHash)) == (tgt ^? _Just . bcHash) then
+            return ()
+        else do
+            rb   <- hashToBlock (cur ^. bcHash . fromDb)
+            prev <- traverse hashToBlock (rb ^? rbContext . bcPrevMain . _Just . fromDb)
+            backfill (prev ^? _Just . rbContext) tgt
+            ExceptT (applyBlock pw rb)
```

Comment:
Let's avoid the use of the stack here and introduce an explicit accumulator for the blocks we need to apply. Would probably also make the code clearer.

Reply:
Let's add a comment emphasizing once more that multiple parallel calls to `applyBlock` are _not_ okay (and that it is the responsibility of the wallet worker to ensure this). It's relevant here because we don't want to have two parallel calls to applyBlock both starting backfilling.

Reply:
I might not have the full context here, but how is a second thread entering `applyBlock`, detecting that and returning (with a `logError` message of course) a "silent failure (even ignoring that fact that we log it)?

Reply:
Because when bugs arise we should not try to deal with them. That way madness (aka Objective-C) lies :)